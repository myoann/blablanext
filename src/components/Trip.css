.trip {
  background-color: #fff;
  border-radius: 16px;
  border: 2px solid #9da1a3;
  transition:
    border-color 0.5s,
    background-color 0.5s;
}

.trip:hover {
  border-color: #00aff5;
  background-color: #e2f8ff;
}

.trip .time,
.trip .addresses {
  transition: 0.5s;
}

.trip:hover .time,
.trip:hover .addresses {
  color: #00aff5;
}

.tripLink {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 150px;
  padding: 16px 24px;
  text-decoration: none;
}

.cheapest {
  font-size: 12px;
  text-transform: uppercase;
  color: #00b894;
}

.vehicle {
  font-size: 12px;
  color: #b2bec3;
  text-transform: capitalize;
}

.timesAndPrice {
  display: flex;
  justify-content: space-between;
}

.timesAndDuration {
  display: flex;
  justify-content: space-between;
  width: 50%;
}

.time {
  font-size: 20px;
  font-weight: 700;
  color: #2d3436;
}

.price {
  font-size: 24px;
  font-weight: 700;
  color: #2d3436;
}

.durationContainer {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 0 10px;
}

.durationLine {
  flex: 1;
  width: 20px;
  border-bottom: 2px solid #a1a9c3;
  margin: 0 10px;
}

.duration {
  color: #a1a9c3;
  font-size: 14px;
}

.addresses {
  display: flex;
  justify-content: space-between;
  width: 50%;
  color: #425486;
  font-size: 14px;
}

.addresses > span:last-child,
.timesAndDuration > span:last-child {
  width: 25%;
}

.transfers {
  width: fit-content;
  padding: 3px 12px;
  border-radius: 50px;
  background-color: rgb(205, 226, 245);
  color: rgb(19, 41, 104);
}
