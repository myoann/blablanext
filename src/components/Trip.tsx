import { ParsedTrip } from "@/types";

import { convertSeconds, currencySymbol } from "@/utils";

import "./Trip.css";
import { useMemo } from "react";

type Props = {
  trip: ParsedTrip;
};

const Trip = ({
  trip: { isCheapest, link, waypoints, duration_in_seconds, price, vehicle },
}: Props) => {
  const firstWaypoint = waypoints[0];
  const lastWaypoint = waypoints[waypoints.length - 1];
  const nbTransfers = waypoints.length - 2;

  const transferText = useMemo(() => {
    if (nbTransfers === 0) {
      return "Direct";
    }

    return `${nbTransfers} ${nbTransfers > 1 ? "Stops" : "Stop"}`;
  }, [nbTransfers]);

  return (
    <li key={link} className="trip">
      <a href={link} target="_blank" className="tripLink">
        {isCheapest && <span className="cheapest">Cheapest</span>}

        <span className="vehicle">
          {vehicle
            ? `${vehicle.make} ${vehicle.model}`
            : "No vehicle information"}
        </span>

        <div className="timesAndPrice">
          <div className="timesAndDuration">
            <span className="time">
              {new Date(firstWaypoint.date_time).toLocaleTimeString("fr-FR", {
                hour: "2-digit",
                minute: "2-digit",
              })}
            </span>

            <div className="durationContainer">
              <div className="durationLine" />
              <div className="duration">
                {convertSeconds(duration_in_seconds)}
              </div>
              <div className="durationLine" />
            </div>

            <span className="time">
              {new Date(lastWaypoint.date_time).toLocaleTimeString("fr-FR", {
                hour: "2-digit",
                minute: "2-digit",
              })}
            </span>
          </div>

          <span className="price">
            {price.amount} {currencySymbol(price.currency)}
          </span>
        </div>

        <div className="addresses">
          <span>
            {firstWaypoint.place.city}, {firstWaypoint.place.country_code}
          </span>
          <span>
            {lastWaypoint.place.city}, {lastWaypoint.place.country_code}
          </span>
        </div>

        <span className="transfers">{transferText}</span>
      </a>
    </li>
  );
};

export default Trip;
