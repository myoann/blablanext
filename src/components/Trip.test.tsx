import { render, screen } from "@testing-library/react";

import Trip from "./Trip";

const defaultProps = {
  trip: {
    link: "link1",
    departure: new Date(2021, 1, 1),
    arrival: new Date(2021, 1, 1),
    waypoints: [
      {
        place: {
          city: "Paris",
          address: "6 rue Menars, 75002 Paris",
          latitude: 48.864716,
          longitude: 2.349014,
          country_code: "FR",
        },
        date_time: "2020-01-10T10:00:00+01:00",
      },
      {
        place: {
          city: "London",
          address: "Charing Cross, London WC2N 5DU",
          latitude: 51.509865,
          longitude: -0.118092,
          country_code: "FR",
        },
        date_time: "2020-01-10T14:00:00+01:00",
      },
    ],
    price: {
      amount: "10",
      currency: "EUR",
    },
    isCheapest: false,
    distance_in_meters: 1000,
    duration_in_seconds: 1000,
  },
};

describe("Trip", () => {
  it("matches the snapshot", () => {
    const { container } = render(<Trip {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it("has a list element containing a link", () => {
    render(<Trip {...defaultProps} />);

    const li = screen.getByRole("listitem");
    expect(li).toBeInTheDocument();

    const link = screen.getByRole("link");
    expect(link).toBeInTheDocument();

    expect(li).toContainElement(link);
  });

  it("has a link which redirects to the trip link", () => {
    render(<Trip {...defaultProps} />);

    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", defaultProps.trip.link);
  });

  it("displays the cheapest text when the trip is the cheapest", () => {
    render(
      <Trip
        {...defaultProps}
        trip={{ ...defaultProps.trip, isCheapest: true }}
      />
    );

    const cheapest = screen.getByText("Cheapest");
    expect(cheapest).toBeInTheDocument();
  });

  it("has a default text when no vehicle is given", () => {
    render(<Trip {...defaultProps} />);

    const vehicle = screen.getByText("No vehicle information");
    expect(vehicle).toBeInTheDocument();
  });

  it("has a text with the vehicle make and model when the vehicle is given", () => {
    render(
      <Trip
        {...defaultProps}
        trip={{
          ...defaultProps.trip,
          vehicle: { make: "Renault", model: "Clio" },
        }}
      />
    );

    const vehicle = screen.getByText("Renault Clio");
    expect(vehicle).toBeInTheDocument();
  });

  it("has a text with the departure time", () => {
    render(<Trip {...defaultProps} />);

    const departureTime = screen.getByText("10:00");
    expect(departureTime).toBeInTheDocument();
  });

  it("has a text with the arrival time", () => {
    render(<Trip {...defaultProps} />);

    const arrivalTime = screen.getByText("14:00");
    expect(arrivalTime).toBeInTheDocument();
  });

  it("has a text with the price", () => {
    render(<Trip {...defaultProps} />);

    const price = screen.getByText("10 â‚¬");
    expect(price).toBeInTheDocument();
  });

  it("has a text with the departure city and country code", () => {
    render(<Trip {...defaultProps} />);

    const departureCity = screen.getByText("Paris, FR");
    expect(departureCity).toBeInTheDocument();
  });

  it("has a text with the arrival city and country code", () => {
    render(<Trip {...defaultProps} />);

    const arrivalCity = screen.getByText("London, FR");
    expect(arrivalCity).toBeInTheDocument();
  });

  it("displays Direct if there is no transfer", () => {
    render(<Trip {...defaultProps} />);

    const transfers = screen.getByText("Direct");
    expect(transfers).toBeInTheDocument();
  });

  it("displays the number of transfers if there is at least one transfer", () => {
    const trip = {
      ...defaultProps.trip,
      waypoints: [
        ...defaultProps.trip.waypoints,
        {
          place: {
            city: "Brussels",
            address: "Brussels",
            latitude: 50.8503463,
            longitude: 4.3517211,
            country_code: "BE",
          },
          date_time: "2020-01-10T12:00:00+01:00",
        },
      ],
    };
    render(<Trip trip={trip} />);

    const transfers = screen.getByText("1 Stop");
    expect(transfers).toBeInTheDocument();
  });
});
