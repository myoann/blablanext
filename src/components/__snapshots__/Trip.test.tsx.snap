// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trip matches the snapshot 1`] = `
<div>
  <li
    class="trip"
  >
    <a
      class="tripLink"
      href="link1"
      target="_blank"
    >
      <span
        class="vehicle"
      >
        No vehicle information
      </span>
      <div
        class="timesAndPrice"
      >
        <div
          class="timesAndDuration"
        >
          <span
            class="time"
          >
            10:00
          </span>
          <div
            class="durationContainer"
          >
            <div
              class="durationLine"
            />
            <div
              class="duration"
            >
              0h16
            </div>
            <div
              class="durationLine"
            />
          </div>
          <span
            class="time"
          >
            14:00
          </span>
        </div>
        <span
          class="price"
        >
          10
           
          â‚¬
        </span>
      </div>
      <div
        class="addresses"
      >
        <span>
          Paris
          , 
          FR
        </span>
        <span>
          London
          , 
          FR
        </span>
      </div>
      <span
        class="transfers"
      >
        Direct
      </span>
    </a>
  </li>
</div>
`;
